### PR TITLE
Fixed conditioning and `scala.MatchError` error bugs

### PR DESCRIPTION
**Pull request by shellshocked2003**
This pull request fixes the following two bugs:
- Conditioning of rock mass based on `minRadius` and `maxAspectRatio` is now working correctly. Issue was that inputs were never passed to `cut` in `RockSlicer`.
- Fixed `scala.MatchError` in `centroid`. Issue was that tolerance in `nonRedundantFaces` was not tight enough so `Face`s that were very close but still outside the block would not be removed. This caused `Face`s with no vertices to be used in centroid calcs.

### CODE REVIEW

File: src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala [hunk 1]
```diff
@@ -328,8 +328,8 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
     linProg.setObjFun(Array(0.0, 0.0, 0.0, 1.0), LinearProgram.MAX)
     faces foreach { face =>
       val coeffs = Array(face.a, face.b, face.c, 1.0)
-      linProg.addConstraint(NumericUtils.applyTolerance(coeffs), LinearProgram.LE,
-                            NumericUtils.applyTolerance(face.d))
+      linProg.addConstraint(coeffs, LinearProgram.LE,
+                            face.d)
```
Comment (jhkolb) [inline src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala]: I know this is picky, but can we just move this up to the previous line?

File: src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala [hunk 2]
```diff
@@ -476,7 +474,7 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
         linProg.addConstraint(faceCoeffs, LinearProgram.LE, rhs)
       }
       val s = linProg.solve().get._2
-      math.abs(s - face.d) <= NumericUtils.EPSILON
+      math.abs(s - face.d) < NumericUtils.EPSILON / 10.0
```
Comment (jhkolb) [inline src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala]: If we're not going to be using `NumericUtils.EPSILON` directly anymore, I would suggest we make this another constant and include it in the `Block` object. That way, if we ever change `NumericUtils.EPSILON`, this will stay the same, and it also seems more clear.

How does that sound to you?

File: src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala [hunk 3]
```diff
@@ -486,12 +484,14 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
     * This function should only be called once all redundant faces have been removed.
     */
   def calcVertices: Map[Face, Seq[Array[Double]]] = {
-    faces.zip (
-      faces map { f1 =>
+    val relevantFaces = nonRedundantFaces
```
Comment (jhkolb) [inline src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala]: Is there a reason for this new `val` definition? Maybe it's left over from your old code where you were filtering by the number of vertices.

File: src/main/scala/edu/berkeley/ce/sparkrocks/RockSlicer.scala [hunk 4]
```diff
@@ -68,7 +68,8 @@ object RockSlicer {
       broadcastJoints.value.zipWithIndex.foldLeft(Seq(seedBlock)) { case (currentBlocks, (joint, idx)) =>
         if (idx % REDUNDANT_ELIM_PERIOD == 0) {
           // When idx is a multiple of REDUNDANT_ELIM_PERIOD, check for geometrically redundant faces
-          currentBlocks.flatMap(_.cut(joint, generation=idx)).map { case block @ Block(center, _, generation) =>
+          currentBlocks.flatMap(_.cut(joint, arguments.minRadius, arguments.maxAspectRatio,
+            generation=idx)).map { case block @ Block(center, _, generation) =>
```
Comment (jhkolb) [inline src/main/scala/edu/berkeley/ce/sparkrocks/RockSlicer.scala]: I think it would be better to write this as

``` scala
currentBlocks.flatMap(_.cut(...)).map { =>
     ...
}
```

File: src/main/scala/edu/berkeley/ce/sparkrocks/SeedJointSelector.scala [hunk 5]
```diff
@@ -74,7 +74,7 @@ object SeedJointSelector {
     } else if (jointSets.head(0).shape.isEmpty) {
       val (seedJoints, leftOverJoints) = searchJointSets(jointSets, inputVolume, numPartitions, jointSetSpan)
       val partitioning = seedJoints.foldLeft(Seq(inputVolume)) { (currentBlocks, joint) =>
-        currentBlocks.flatMap(_.cut(joint))
+        currentBlocks.flatMap(_.cut(joint, minSize = 0.0, maxAspectRatio = Double.PositiveInfinity))
```
Comment (jhkolb) [inline src/main/scala/edu/berkeley/ce/sparkrocks/SeedJointSelector.scala]: Now that I think about it, why did we decide we couldn't give `minSize` and `maxAspectRatio` default values?

It seems like 0 and `Double.PositiveInfinity` are such a common use case, that it would still be valuable to have them as defaults, and we can just specify them explicitly in `RockSlicer`.

File: src/test/scala/edu/berkeley/ce/sparkrocks/BlockSpec.scala [hunk 6]
```diff
@@ -394,7 +394,6 @@ class BlockSpec extends FunSuite {
     val block = Block(Array(1.0, 1.0, 1.0), List(face1, face2, face3))
     val expectedIntersection = Map(
       face1 -> List.empty[Array[Double]],
-      face2 -> List.empty[Array[Double]],
```
Comment (jhkolb) [inline src/test/scala/edu/berkeley/ce/sparkrocks/BlockSpec.scala:397]: Why was this line removed?

**Review by jhkolb — CHANGES_REQUESTED** [review]
Just a few small things to look at.

File: src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala (hunk 1 above)
Comment (shellshocked2003) [reply src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala]: Sure :)

File: src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala (hunk 2 above)
Comment (shellshocked2003) [reply src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala]: That sounds reasonable to me. It's also more explicit about where were using something different.

File: src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala (hunk 3 above)
Comment (shellshocked2003) [reply src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala]: Oops. Yep, leftovers. It's gone now.

File: src/main/scala/edu/berkeley/ce/sparkrocks/RockSlicer.scala (hunk 4 above)
Comment (shellshocked2003) [reply src/main/scala/edu/berkeley/ce/sparkrocks/RockSlicer.scala]: You mean just in terms of formatting or am I missing something? I've changed it based on what I think you mean, but let me know if I misinterpreted!

File: src/main/scala/edu/berkeley/ce/sparkrocks/SeedJointSelector.scala (hunk 5 above)
Comment (shellshocked2003) [reply src/main/scala/edu/berkeley/ce/sparkrocks/SeedJointSelector.scala]: I think the reason we decided not to go with it was because we thought we'd have to bind one of the arguments and do some overly fancy stuff.

Looking at it now, your suggestions seems like a good idea. I'll change it.

File: src/test/scala/edu/berkeley/ce/sparkrocks/BlockSpec.scala (hunk 6 above)
Comment (shellshocked2003) [reply src/test/scala/edu/berkeley/ce/sparkrocks/BlockSpec.scala:397]: This was due to the extra `val` that you'd mentioned earlier. `face2` will be removed if you call `nonRedundantFaces`. I added it back in since the extra `val` that triggered a call to `nonRedundantFaces` is now gone.

**Comment by shellshocked2003** [comment]
I've addressed all your comments, but take a look and see what you think. There was one that I wasn't entirely sure about.

File: src/main/scala/edu/berkeley/ce/sparkrocks/SeedJointSelector.scala (hunk 5 above)
Comment (jhkolb) [reply src/main/scala/edu/berkeley/ce/sparkrocks/SeedJointSelector.scala]: Cool!

What this means is that you  now don't have to specify the `minSize` and `maxAspectRatio` explicitly anymore if you're sticking to the defaults. So I think there are a few places where you can remove those arguments in the code.

Comment (shellshocked2003) [reply src/main/scala/edu/berkeley/ce/sparkrocks/SeedJointSelector.scala]: Oops, sorry! Thought I got them all, but the ones in `SeedJointSelector` slipped past me... They're gone now.

**Review by jhkolb — APPROVED** [review]

**Comment by jhkolb** [comment]
Looks good, just merged this in.
